### ISSUE TITLE
Manually specifying Range HTTP header

### ISSUE THREAD

**Issue by AleAmla987** [google/ExoPlayer#9323]
Hi there! I am building an application with a list of ogg audios and they weigh about 30mb each, I would like to know how I can add the header Range to it so that it loads faster. How can I make the rest of the audio on my list also load a maximum of seconds so that when I go from one audio to another it plays immediately.

**Comment by icbaker** [comment]
I'm afraid I don't really understand what you're asking.

> I would like to know how I can add the header Range to it so that it loads faster.

If you're talking about the [HTTP `Range`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) header then ExoPlayer's 3 official HTTP datasources will all add this header when requesting data from a non-zero start position. You shouldn't need to do anything. If you see this not happening then please provide clear repro instructions and we'll investigate.

> How can I make the rest of the audio on my list also load a maximum of seconds so that when I go from one audio to another it plays immediately.

If I'm understanding correctly this sounds like a duplicate of https://github.com/google/ExoPlayer/issues/3327

**Comment by AleAmla987** [comment]
> I'm afraid I don't really understand what you're asking.
> 
> > I would like to know how I can add the header Range to it so that it loads faster.
> 
> If you're talking about the [HTTP `Range`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) header then ExoPlayer's 3 official HTTP datasources will all add this header when requesting data from a non-zero start position. You shouldn't need to do anything. If you see this not happening then please provide clear repro instructions and we'll investigate.
> 
> > How can I make the rest of the audio on my list also load a maximum of seconds so that when I go from one audio to another it plays immediately.
> 
> If I'm understanding correctly this sounds like a duplicate of #3327

I apologize, I am Spanish-speaking and my English is not very good. I am creating an audio book app and testing with postman I discovered that if I add Range byte = 0-500 in the header the response is almost immediate, I would like to know if this is possible to use in the player so that the audio plays faster.

**Comment by icbaker** [comment]
I think we need more information to understand the details of what you're trying to achieve.

Some possibly relevant points (apologies if any of these are obvious):
* ExoPlayer will start playing content as soon as it receives enough bytes. The exact number of bytes needed to start playback depends on the structure of the file, but generally this means receiving all the metadata at the top of the file and at least one 'sample' of actual compressed content data.
  * ExoPlayer **does not** wait for the entire file to be downloaded before starting playback.
* I don't believe that specifying the `Range` HTTP header would allow, in general, an HTTP server to respond faster ([time to first byte](https://en.wikipedia.org/wiki/Time_to_first_byte)).
  * Whether you're requesting a whole 600GB file or just the first 500 bytes of it, it should be able to respond with the first 500 bytes just as quickly in either case.
  * Obviously it **will** speed up the time to finishing the download (because you're only fetching 500 bytes instead of the whole file). But as above, the time to download the whole file isn't relevant to how long it takes ExoPlayer to **start** playback. 
  * It's not clear from your comment whether your postman experiments are measuring the time-to-first-byte or time to complete the response. If you're observing that your HTTP server **starts** responding faster when the `Range:` header is set, then that's something you might want to investigate on the server side.

So it's not quite clear to me exactly what you're aiming to speed up by manually specifying this header. That said, if you want to try specifying it anyway, you should be able to do it by modifying the `DataSpec` on the fly using a [`ResolvingDataSource`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/upstream/ResolvingDataSource.html). To affect the values specified in the `Range:` header you will want to set/modify `DataSpec#position` and `DataSpec#length`. Obviously if you change this to arbitrarily request only the first 500 bytes of a file then playback will almost certainly stop early, and may well fail (because the resulting file may not be valid). The exact behaviour depends on the format of the file in question.
